---
author: "Bernd Klaus"
bibliography: MA_end_to_end.bib
csl: biometrics.csl
output: word_document
vignette: "%\\VignetteIndexEntry{An end to tend workflow for differential gene expression
  \ \n"
  using Affymetrix microarrays} %\\VignetteEngine{knitr::rmarkdown} %\\VignetteEncoding{UTF-8}
---


<!--
To compile this document run
graphics.off();rm(list=ls());rmarkdown::render('MA-Workflow.Rmd');
-->

<!--
F1000 manuscript compilation
graphics.off();rm(list=ls());rmarkdown::render('MA-Workflow.Rmd', 
c("html_document", "pdf_document"), clean=FALSE);

produce tex on command line

pandoc MA-Workflow.utf8.md --output MA-Workflow.tex --to latex \
--from markdown+autolink_bare_uris+ascii_identifiers+tex_math_single_backslash-implicit_figures \
--template /home/bernd/R/x86_64-pc-linux-gnu-library/3.3/rmarkdown/rmd/latex/default.tex \
--highlight-style tango --latex-engine pdflatex --variable 'geometry:margin=1in' \
--bibliography MA_end_to_end.bib --biblatex

sed 's/\\autocite/\\cite/g' MA-Workflow.tex | \
sed 's/\\(/$/g' | \
sed 's/\\)/$/g' | \
sed 's/\.png/\.pdf/g' | \
sed 's/MA-Workflow_files\/figure-latex\///g' > MA-Workflow_clean.tex
-->

<!--
     # a list of all required libraries:
     reqlibs = sub(".*library\\(\"(.*?)\"\\).*","\\1",grep("library\\(",readLines("MA-Workflow.Rmd"),value=TRUE))
     find.package(reqlibs)
-->

<!--
<script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>
-->

#一个从原始数据到富集分析的芯片分析流程
jixing475

```{r options, include=FALSE}
library(BiocStyle)
library(knitr)
options(digits=3, width=80)
opts_chunk$set(echo=TRUE,tidy=FALSE,include=TRUE,
               dev=c('png', 'pdf'), fig.width = 6, fig.height = 3.5,
               comment = '  ', dpi = 300,
cache = TRUE)
```


## Abstract

In this article, we walk through an end--to--end Affymetrix microarray differential
expression workflow using Bioconductor packages. This workflow is directly 
applicable to current "Gene" type arrays, e.g. the HuGene or MoGene arrays
but can easily adapted to  similar platforms.
The data re--analyzed is a typical clinical microarray data set that 
compares inflammed and non--inflammed colon tissue in two disease subtypes.
We will start from the raw data CEL files, show how to import them into a 
Bioconductor ExpressionSet, perform quality control and normalization and finally
differential gene expression (DE) analysis, followed by some enrichment 
analysis. As experimental designs can be complex, a self contained introduction 
to linear models is also part of the workflow. 

## Introduction

In this article we introduce a complete workflow for a typical (Affymetrix) microarray
analysis. Data import, preprocessing, differential expression  and
enrichment analysis are discussed. We also introduce some necessary mathematical
background on linear models along the way.

Our aim is to analyze differential expression (DE) between the tissues in the two  diseases.

## Required packages and other preparations
```{r}
library(Biobase)
library(oligoClasses)
library(knitr)
library(BiocStyle)
library(oligo)
library(geneplotter)
library(arrayQualityMetrics)
library(ggplot2)
library(dplyr)
library(LSD)
library(gplots)
library(RColorBrewer)
library(ArrayExpress)
library(arrayQualityMetrics)
library(stringr)
library(matrixStats)
library(topGO)
library(genefilter)
library(pd.hugene.1.0.st.v1)
library(pd.hg.u133a.2)
library(hugene10sttranscriptcluster.db)
library(pheatmap)
library(mvtnorm)
library(DAAG)
library(multcomp)
library(limma)
library(ReactomePA)
library(clusterProfiler)
library(openxlsx)
library(devtools)
library(biomaRt)
library(EnrichmentBrowser)
set.seed(777)
raw_data_dir <- file.path(getwd(), "rawDataMAWorkflow")
```


#下载数据
##从公共数据库下载原始数据
Download the raw data and the annotation data 

With the code below, we download the raw data from
[ArrayExpress](https://www.ebi.ac.uk/arrayexpress/) [@Kolesnikov_2014].
It is saved in the directory **raw\_data\_dir** which defaults to the subdirectory
`rawDataMAWorkflow` of  the current working directory. The names 
of the downloaded files are returned as a list.


The first step of the analysis is to download the raw data CEL files. These files
are produced by the array scanner software and contain the probe intensities 
measured. The data have been deposited at [ArrayExpress](https://www.ebi.ac.uk/arrayexpress/)
and have the accession code **E-MTAB-2967** **E-GEOD-30528"**. 

Each ArrayExpress data set has a landing page summarizing the data set, 
and we use the `r Biocpkg("ArrayExpress") ` Bioconductor package to obtain the ftp 
links  to the raw data files ([Data from Palmieri et. al. on ArrayEpress](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-2967/)).

Download the raw data from from ArrayExpress
```{r}
#if(!dir.exists(raw_data_dir)){
#    dir.create(raw_data_dir)
#}
#
#anno_AE <- getAE("E-GEOD-30528", path=raw_data_dir, type="raw")
```


## 获取芯片的实验设计信息
Information stored in ArrayExpress

Each dataset at ArrayExpress is stored according to the MAGE--TAB 
(MicroArray Gene Expression Tabular) specifications as a collection of
tables bundled with the raw data. The MAGE--TAB format specifies  up to five 
different types  of files, namely the
Investigation Description Format (IDF), the Array Design Format (ADF),
the Sample and Data Relationship Format (SDRF), the raw data files and
the processed data files.

For use, the IDF and the **SDRF** file are important. 
The IDF file contains top level information
about the experiment including title, description, submitter contact details and
protocols. The SDRF file contains essential information on the experimental 
samples, e.g. the experimental group(s) they belong to. 


We now import the SDRF file directly from ArrayExpress
in order to obtain the sample annotation.

The raw data consists of one CEL file per sample (see below) and
we use the CEL file names as row names for the imported data. 
These names are given in a column
named `Array.Data.File` in the SDRF table. We turn the SDRF table into an
` AnnotatedDataFrame ` from the `r Biocpkg("Biobase") ` package that we
will need later to create an `ExpressionSet` for our data [@Bioc].


```{r getSDRF}
SDRF <- read.delim(
url("http://www.ebi.ac.uk/arrayexpress/files/E-GEOD-30528/E-GEOD-30528.sdrf.txt"))

rownames(SDRF) <- SDRF$Array.Data.File
SDRF <- AnnotatedDataFrame(SDRF)
```


## Bioconductor ExpressionSets

The data in an ExpressionSet consist of

+ **assayData**: Expression data from microarray experiments.

+ **metaData**: A description of the samples in the experiment
(phenoData), metadata about the features on the chip or technology used for the
experiment (featureData), and further annotations for the features, for example
gene annotations from biomedical databases (annotation).

+ **experimentData**: A flexible structure to describe the experiment.

The ExpressionSet class coordinates all of these data, so that one does not 
have to worry about the details. However, some constrains have to be met. 
In particular, the rownames of the `phenoData` (which holds the content of the
SDRF file) have to match the column names of the assay data (as they represent the sample 
identifiers), while the row names of the expression data have to match the 
row names of the `featureData` (as they represent the feature identifiers). This is 
illustrated in the figure.

```{r sumexp, echo=FALSE, fig.show="asis"}
par(mar=c(0,0,0,0))
plot(1,1,xlim=c(0,100),ylim=c(0,100),bty="n",
     type="n",xlab="",ylab="",xaxt="n",yaxt="n")
polygon(c(45,80,80,45),c(10,10,70,70),col=rgb(1,0,0,.5),border=NA)
polygon(c(45,80,80,45),c(68,68,70,70),col=rgb(1,0,0,.5),border=NA)
text(62.5,40,"assay(s)", cex = 1)
text(62.5,30,"e.g. 'exprs'", cex = 1)
polygon(c(20,40,40,20),c(10,10,70,70),col=rgb(0,0,1,.5),border=NA)
polygon(c(20,40,40,20),c(68,68,70,70),col=rgb(0,0,1,.5),border=NA)
text(30,40,"featureData", cex = 1)
polygon(c(45,80,80,45),c(75,75,90,90),col=rgb(.5,0,.5,.5),border=NA)
polygon(c(45,47,47,45),c(75,75,90,90),col=rgb(.5,0,.5,.5),border=NA)
text(62.5,82.5,"phenoData", cex = 1)
```

You can use the functions ` pData ` and ` fData ` to extract
the sample and feature annotation respectively from an ` ExpressionSet `.
The function ` exprs ` will return the expression data itself as a matrix.

## 读取原始数据
Import of the raw microarray data 

The analysis of Affymetrix arrays starts with CEL files. These are the result
of the processing of the raw image files using the Affymetrix software and contain
estimated probe intensity values. Each CEL file additionally contains some 
metadata, such as a chip identifier. 

The function ` read.celfiles `
from the `r Biocpkg("oligo") ` [@oligo] can be
used to import the files. The package automatically uses
`r  Biocannopkg("pd.hugene.1.0.st.v1") ` as the chip annotation package as the 
chip--type is also stored in the .CEL files.

We specify our `AnnotatedDataFrame` created earlier as `phenoData`. Thus, We have 
to be sure that we import the CEL files in the order that corresponds to the SDRF
table --- to enforce this, we use the column `Array.Data.File` of the `SDRF` table as the `filenames` 
argument.

Finally, we check whether the object created is valid. (e.g. sample names 
match between the different tables).

We collect the information about 
the CEL files and import and them into the  variable `raw_data`:


```{r importCelfiles, results="hide", eval=TRUE, dependson="getSDRF", warning = FALSE }
raw_data <- read.celfiles(filenames = file.path(raw_data_dir, 
                                                SDRF$Array.Data.File),
                         verbose = FALSE, phenoData = SDRF)
validObject(raw_data)
```

##保留phenoData中需要的列
We now inspect the raw data a bit and retain only those columns that are 
related to the experimental factors of interest (identifiers of the individuals, 
disease of the individual and the mucosa type).

```{r inspectPhenoData, eval=TRUE }
head(pData(raw_data))
head(exprs(raw_data))
stopifnot(validObject(raw_data))

pData(raw_data) <- pData(raw_data)[, c("Source.Name",
                                     "Characteristics.disease.state.")]
```


# 对原始数据做质量检控
Quality control of the raw data

##先看看数据符合预期 PCA
The first step after the intial data import is the quality control of the data.
Here we check for outliers and try to see whether the data clusters as expected, 
e.g. by the experimental conditions. We use the identifiers of the individuals
as plotting symbols.

The PCA (performed on the log--intensity scale) plot of the raw data shows 
that the first principal component differentiates  between the diseases. 
However, the intensity boxplots  show that the  intensity distributions 
of the individual arrays are quite different,
indicating the need of an appropriate normalization, which we will discuss next.
```{r quality_control_raw_data}
exp_raw <- log2(exprs(raw_data))
PCA_raw <- prcomp(t(exp_raw), scale = FALSE)

dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                    Disease = pData(raw_data)$Characteristics.disease.state.)
        
(qplot(PC1, PC2, data = dataGG, color =  Disease,
       main = "PCA plot of the raw data (log-transformed)", size = I(2), 
       asp = 1.0, geom = "text",
       label = Disease)
     + scale_colour_brewer(palette = "Set2"))

boxplot(raw_data, target = "core", 
        main = "Boxplots of log2-intensities for the raw data")
```


## 质量检测的R packages **arrayQualityMetrics**
A wide range of quality control plots can be created using the package 
`r Biocpkg("arrayQualityMetrics") ` [@AQM]. The package  produces an html report, 
containing the quality control plots together with a description of their
aims and an identification of possible outliers. We don't discuss this tool in detail
here, but the code below can be used to create a report for our raw data. 

**注意**这块代码在R 3.3.1的情况下才成功
```{r arrayQualityMetricsRaw, eval = FALSE}
dev.off(which(unlist(.Devices) == "gridsvg"))
arrayQualityMetrics(expressionset = raw_data,
    outdir = "Report_for_DKD_raw",
    force = TRUE, do.logtransform = TRUE,
    intgroup = "Characteristics.disease.state.")
```

# 数据标准化处理
Background adjustment, calibration, summarization and annotation

## Background adjustment

After the initial import and quality assessment, the next step in processing of
microarray data is background adjustment. This  is essential because a part of the
measured probe intensities are due to non-specific hybridization and the noise 
in the optical detection system. Therefore, observed intensities need to be adjusted to 
give accurate measurements of specific hybridization.

## Across--array normalization (calibration)

Without proper normalization across arrays, it is impossible to compare measurements from
different array hybridizations due to many obscuring sources of variation.
These include different efficiencies of
reverse transcription, labeling or hybridization
reactions, physical problems with the arrays, reagent batch effects, and laboratory
conditions. 

## Summarization

After normalization, summarization is needed because on the Affymetrix 
platform transcripts are represented
by multiple probes. For each gene, the background adjusted and normalized intensities
need to be summarized into one quantity that estimates an amount proportional to
the amount of RNA transcript.

After the summarization step, the summarized data can be annotated with various
information, e.g. gene symbols and EMSEMBL gene identifiers. There is an
annotation database available from Bioconductor 
for our platform, namely the package
`r Biocannopkg("hugene10sttranscriptcluster.db") `.

You can view its content like this

```{r annotation data base content, eval = TRUE}
#head(ls("package:hugene10sttranscriptcluster.db"))
```

## 关于老芯片和新的芯片
Old and new "probesets" of  Affymetrix microarrays

Traditionally, Affymetrix arrays (the so--called 3' IVT arrays)
were probeset based: a certain fixed group of probes were part of a probeset
which represented a certain gene or transcript (note however, that a
gene can be represented by multiple probesets).

The more recent "Gene" and "Exon" Affymetrix arrays are exon based and hence there 
are two levels of summarization. The exon level summarization 
leads to "probeset" summary. However,
these probesets are not the same as the  probesets of the previous
chips, which usually represented a gene/transcript. Furthermore, there
are also no longer designated match/mismatch probes present on "Gene" type chips.

For the newer Affymetrix chips a gene/transcript level summary is given 
by "transcriptct clusters". Hence the appropriate annotation package is called
`r Biocpkg("hugene10sttranscriptcluster.db") `.

To complicate things even a bit more, note that the "Gene"  arrays  were created
as affordable versions of the "Exon" arrays by taking the "good"  probes from the Exon
array. So the notion of a probeset is based on the
original construction of the probesets on the Exon array, which contains
usually at least four probes.

But since Affymetrix selected only a the subset of "good" probes for the 
Gene arrays, a lot of the probesets on the "Gene" arrays are made up of three 
or fewer probes. Thus, a summarization on the probeset / exon level 
is not recommended for "Gene" arrays but nonetheless possible by using the 
`r Biocannopkg("hugene10stprobeset.db") ` annotation package.

## One--go preprocessing in oligo

The package `r Biocpkg("oligo") ` allows us to perform background correction, 
normalization and summarization in one single step using a deconvolution 
method for background correction, quantile normalization and
the RMA (robust multichip average) algorithm for summarization.

This series of steps as a whole is commonly referred to as RMA algorithm,
although strictly speaking RMA is merely a summarization method
[@Irizarry_2003; @Bolstad_2003; @Irizarry_2003a].

```{r RMAcalibration, eval=TRUE}
palmieri_eset <- oligo::rma(raw_data)
```

The parameter `target` defines the degree of summarization, the
default option of which is "core", using transcript clusters containing
"safely" annotated genes. Other options for `target` include "extended"
and "full". For summarization on the exon level (not recommended for Gene
arrays), one can use "probeset" as the target option.

Although other methods for background correction and normalization exist,
RMA is usually a good default choice.
RMA shares information across arrays and
uses the versatile quantile normalization method  that
will make the array intensity distributions match. However, it is preferable 
to apply it only after outliers have been removed. 
The quantile normalization algorithm used by RMA
works by replacing values by the average of identically
ranked (with a single chip) values across arrays. A more detailed 
description can be found on the [Wikipedia page](https://en.wikipedia.org/wiki/Quantile_normalization) 
about it.

An alternative to quantile normalization is the `r Biocpkg("vsn") ` algorithm,
that performs background correction and normalization by robustly
shifting and scaling log--scale intensity values within arrays [@vsn]. This is
less "severe" than quantile normalization. 

## 对标准化后的数据进行质检
Quality assessment of the calibrated data

We now produce a clustering and another PCA plot using the calibrated data. In 
order to display a heatmap of the sample--to--sample distances, we first compute the distances using
the `dist` function. We need to transpose the expression values since
the function computes the distances between the rows (i.e. genes in our case) by
default. The default distance  is the Euclidean one. However this can  be
changed and we choose the manhatten distance here (it uses absolute instead of
squared distances). We set the diagonal of the distance matrix to ` NA ` in order 
to increase the contrast of the color coding. Those diagonal entries do 
not contain information since the distance of a sample to itself is
always equal to zero.



```{r PCAMetricsCalibrated, eval = TRUE }
exp_palmieri <- exprs(palmieri_eset)
PCA <- prcomp(t(exp_palmieri), scale = FALSE)

dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    Disease = pData(palmieri_eset)$Characteristics.disease.state.)
        
(qplot(PC1, PC2, data = dataGG, color =  Disease,
       main = "PCA plot of the calibrated data", size = I(2), asp = 1.0)
       + scale_colour_brewer(palette = "Set2"))
```
```{r PCAMetricsCalibrated_2, fig.height = 8.5, eval = TRUE }
dists <- as.matrix(dist(t(exp_palmieri), method = "manhattan"))
colnames(dists) <- NULL
diag(dists) <- NA
rownames(dists) <-  pData(palmieri_eset)$Characteristics.disease.state.
hmcol <- colorRampPalette(rev(brewer.pal(9, "PuOr")))(255)

pheatmap(dists, col = rev(hmcol), clustering_distance_rows = "manhattan",
                                 clustering_distance_cols = "manhattan")
```


## 过滤低杂交密度的探针
Filtering based on intensity 

We now filter out lowly expressed genes. Microarray data commonly show a large number
of probes in the background intensity range. They also do not change much across arrays. 
Hence they combine a low variance with a low intensity. Thus, they could end up being detected 
as differentially expressed although they are barely above the "detection" limit 
and are not very informative in general. We will perform
a "soft" intensity based filtering here, since this is recommended by `r Biocpkg("limma")`'s 
[@limma; @Smyth_2004] user guide (a package we will use below for the differential 
expression analysis). However, note that 
a variance based filter might exclude a similar
set of probes in practice. In the histogram of the gene--wise medians, we 
can clearly see an enrichment of low medians on the left hand side. 
These represent the genes we want to filter. 

In order to infer a cutoff from the data, we inspect the histogram of the median--intensities.
We visually fit a central normal distribution given by \(0.5 \cdot N(5.1, 1.18)\) 
to the probe--wise medians, which represents their typical behavior in the data set at hand.

Then we use the 5% quantile of this distribution as a threshold,  We keep
only those genes that show an expression higher than the threshold in at least
as many arrays as in  the smallest experimental group.

```{r expGroups, dependson="PCAMetricsCalibrated"}
#no_of_samples <- table(paste0(pData(palmieri_eset)$Factor.Value.disease., "_", 
#                        pData(palmieri_eset)$Factor.Value.phenotype.))
#no_of_samples 
```

In our case this would be r min(no_of_samples).


```{r intensityBasedFiltering, fig.width=10, fig.height=6, eval=TRUE}
#palmieri_medians <- rowMedians(exprs(palmieri_eset))
#
#hist_res <- hist(palmieri_medians, 100, col="#e7efd8", freq = FALSE, 
#            main = "Histogram of the median intensities", 
#            xlab = "Median intensities")
#
#emp_mu <- hist_res$breaks[which.max(hist_res$density)]
#emp_sd <- mad(palmieri_medians)/2
#prop_cental <- 0.50
#
#lines(sort(palmieri_medians), prop_cental*dnorm(sort(palmieri_medians),
#                              mean = emp_mu , sd = emp_sd),
#                              col = "grey10", lwd = 4)
#
#cut_val <- 0.05 / prop_cental
#thresh_median <- qnorm(0.05 / prop_cental, emp_mu, emp_sd)
#
#samples_cutoff <- min(no_of_samples)
#
#idx_thresh_median <- apply(exprs(palmieri_eset), 1, function(x){
#                                sum(x > thresh_median) >= samples_cutoff})
#table(idx_thresh_median)
#
#palmieri_filtered <- subset(palmieri_eset, idx_thresh_median)
```


## 探针注释 
Annotation of the transcript clusters

Before we continue with the linear models for microarrays and differential
expression  we  describe how to add "feature Data", i.e. annotation
information to the transcript cluster identifiers stored in the featureData of
our ExpressionSet. We use the function `select`
from `r Biocpkg("AnnotationDbi") ` to query the gene symbols and associated
short descriptions for the transcript clusters. For each cluster, we add the 
gene symbol and a short description of the gene the cluster represents. 


```{r annotateData, eval=TRUE, dependson="intensityBasedFiltering", message = FALSE}
affydb<- annPkgName(palmieri_eset@annotation,type = "db")
#library(affydb,character.only = TRUE)
library("hgu133a2.db",character.only = TRUE)
anno_palmieri  <- AnnotationDbi::select(hgu133a2.db,
                                  keys=(featureNames(palmieri_eset)),
                                  columns = c("SYMBOL", "GENENAME"),
                                  keytype="PROBEID")

```

## 删除多重映射
Removing multiple mappings and building custom annotations

Many transcript--cluster identifiers will map to multiple gene symbols.
We compute a summary table in the code below to see how many there are.

```{r multipleMappings, dependson="annotateData"}
probe_stats <- anno_palmieri   %>%
    group_by(PROBEID) %>%
    summarize(no_of_matches = n_distinct(SYMBOL)) %>%
    filter(no_of_matches > 1)

probe_stats

dim(probe_stats)
```

We have  over 2000 transcript--clusters that map to multiple gene symbols.
It is difficult to decide which mapping is "correct". Therefore,
we exclude these transcript--clusters. Additionally, we also exclude
transcript--clusters that do not map to gene symbols.


```{r excludeMultipleMappings, dependson="multipleMappings", cache=TRUE}
ids_to_exlude <- ((featureNames(palmieri_eset) %in% probe_stats$PROBEID) |
               featureNames(palmieri_eset)  %in% subset(anno_palmieri ,
                                                           is.na(SYMBOL))$PROBEID)
table(ids_to_exlude)

palmieri_final <- subset(palmieri_eset, !ids_to_exlude)

validObject(palmieri_final)

fData(palmieri_final)$PROBEID <- rownames(fData(palmieri_final))
fData(palmieri_final) <- left_join(fData(palmieri_final), anno_palmieri)

# restore rownames after left_join
rownames(fData(palmieri_final)) <-fData(palmieri_final)$PROBEID 
    
validObject(palmieri_final)
```

Alternatively, one can re--map the probes of the array
to a current annotation, a workflow to do this for Illumina arrays is given in
@Arloth_2015.
Essentially, the individual probe sequences are re--aligned to an in--silico
"exome" that consists of all annotated transcript exons.

##建造自己的注释数据库
In any case, the package `r Biocpkg("pdInfoBuilder") ` can be used to build custom
annotation packages for use with `r Biocpkg("oligo") `. In order to do this,
PGF / CLF files  (called "Library files" on the Affymetrix website) as well
as the probeset annotations are required. The probesets typically represent
a small stretches of the genome (such as a single exon) and multiple probesets
are then used to form a transcript cluster.

The CLF file contains information about the location of
individual probes on the array. The PGF file then contains the individual probe
sequences and shows the probeset they belong to. Finally, The probeset annotation .csv
then contains information about which probesets are used
in which transcript cluster. Commonly, multiple probesets are used in one
transcript cluster and some probesets are contained in multiple transcript
clusters.



## 线性模型
A linear model for the data

The original paper is interested in changes in transcription that occur
between inflamed and adjacent non--inflamed mucosal areas of the colon.
This is studied in both inflammatory bowel disease types.

Since we have two arrays per individual, the first factor we need
is a blocking factor for the individuals that will absorb differences between
them. Then we create a factors that give us the grouping for the diseases and
the tissue types. We furthermore simplify the names of the
diseases to UC and DC, respectively. Then, we create two design matrices, 
one for each of the two diseases
as we will analyze them separately in order to follow the  analysis 
strategy of the original paper closely (one could also fit a joint model to 
the complete data set, however, the two diseases might behave very differently 
so that a joint fit might not be appropriate).

##设计矩阵的构建
```{r createDesign, eval=TRUE, dependson="excludeMultipleMappings" }
disease <- str_replace_all(pData(palmieri_final)$Characteristics.disease.state.,
                           " ","_")
disease <- factor(ifelse(disease=="control","control","DKD"))
design <- model.matrix(~-1+disease)

```

##检验设计矩阵的秩
We can inspect the design matrices and test their rank.

```{r inspectDesignMatrix, eval = TRUE, dependson="createDesign"}
head(design)
dim(design)
min(svd(design)$d)
```


## 构建实验组和对照组的比较
Contrasts and hypotheses tests

We  now fit the linear models and
define appropriate contrasts to test hypotheses of interest.
We want to compare the inflamed  to the the non--inflamed tissue.
Thus, we create a contrast matrix consisting of one row. `r Biocpkg("limma")` 's function
`makeContrasts` creates this matrix from a synbolic description of the contrast of
interest. We can fit the linear model, compute the moderated \(t\)--statistics
by calling the ` eBayes ` function and finally extract the number of
differentially expressed genes while controlling the FDR by
requiring BH--corrected p--value below a certain threshold.

```{r createContrastMatrixAndFitModel, eval=TRUE, dependson="createDesign" }
contrast_matrix <- makeContrasts(contrasts = "diseaseDKD - diseasecontrol",levels = design)

palmieri_fit_DKD <- eBayes(contrasts.fit(lmFit(palmieri_final,
                                design = design),
                                contrast_matrix))
```

## 提取假设检验的比较结果
Extracting results

Results can be extracted by use of the `topTable` function. We extract
the comparisons for both Crohn's disease  as well as ulcerative colitis and
sort the results by their absolute \(t\)--statistics. As a diagnostic check, we also
plot the  p--value histogram: We expect a uniform distribution for the
p--values that correspond to true null hypotheses, while the a peak near zero
shows a enrichment for low p--values corresponding to differentially expressed (DE)
genes.  A p--value less than 0.001 was used in the original paper as a significance
cutoff leading to 298 (CD) and 520 (UC) DE--genes for the two diseases.

We call around 500/1000 genes in the two conditions at the same cutoff, this 
higher number of DE genes identified is probably
due to the increased power from the blocking according to the individuals
and the moderated variance estimation that `r Biocpkg("limma") ` performs.

##提取结果，并看 p value 的分布
```{r extractResults, eval = TRUE, dependson="createContrastMatrixAndFitModel", message=FALSE}
table_CD <-  topTable(palmieri_fit_DKD, number = Inf,
                      coef ="diseaseDKD - diseasecontrol")
head(table_CD)

table(table_CD$adj.P.Val < 0.05)

table(table_CD$P.Value < 0.001)

hist(table_CD$P.Value, col = brewer.pal(3, name = "Set2")[1],
     main = "disease_DKD - disease_control", xlab = "p-values")
```




#  GO 注释
Gene ontology (GO) based enrichment analysis

We can now try characterize the identified differentially expressed genes
a bit better by performing an GO enrichment analysis. Essentially the
gene ontology ([http://www.geneontology.org/](http://www.geneontology.org/)) is 
a hierarchically organized
collection of functional gene sets [@Ashburner_2000, @GO_2015, @du_Plessis_2011]. 


## 看看背景基因和差异基因是否有相似的表达强度分布（最好是有）
Matching the background set of genes

The function ` genefinder ` from the `r Biocpkg("genefilter") ` [@Bourgon_2010] 
will be used to find a background set of genes that are similar in expression 
to the differentially expressed genes. We then check whether 
the background has roughly the same distribution
of average expression strength as the foreground.

We do this in order not to select a biased background since the gene set testing
is performed by a simple Fisher test on a 2x2 table. Note that this approach
is very similar to commonly used web tools like **GOrilla** [@Eden_2009]. 
Here we focus on the  CD subset of the  data.

For every differentially expressed gene, we try to find genes with similar
expression.

```{r GOAnalysisCreateBackgrounds, eval=TRUE, dependson=c("extractResults", "excludeMultipleMappings"), warning=FALSE, message=FALSE}
DE_genes_CD <- subset(table_CD, adj.P.Val < 0.1)$PROBEID

back_genes_idx <- genefinder(palmieri_final, as.character(DE_genes_CD), 
                       method="manhattan", scale="none")

back_genes_idx <- sapply(back_genes_idx, function(x)x$indices)

back_genes <-featureNames(palmieri_final)[back_genes_idx]
back_genes <- setdiff(back_genes, DE_genes_CD)

    
intersect(back_genes, DE_genes_CD)
length(back_genes)

multidensity(list(
        all=  table_CD[,"AveExpr"] ,
        fore= table_CD[DE_genes_CD , "AveExpr"],
        back= table_CD[rownames(table_CD) %in% back_genes, "AveExpr"]),
        col = c("#e46981", "#ae7ee2", "#a7ad4a"),
     xlab="mean expression",
   main = "DE genes for CD - background - matching")
```

We can see that the matching returned a sensible result and can now
perform the actual testing. For this purpose we use the `r Biocpkg("topGO") ` which
implements a nice interface to Fisher testing and also has additional algorithms
taking the GO structure into account, by e.g. only reporting the most specific
gene set in the hierarchy [@Alexa_2006].

The GO has three top ontologies, cellular component (CC), biological  processes
(BP), and molecular function (MF). For illustrative purposes we limit ourselves 
to the BP category here. 

## GO 分析
Running topGO

We first create a factor `all_genes` which indicates for every gene in
our background / universe, whether it is differentially expressed or not.

```{r createFactorOfInterestingGenes, dependson="GOAnalysisCreateBackgrounds", eval=TRUE}
gene_IDs <- rownames(table_CD)
in_universe <- gene_IDs %in% c(DE_genes_CD ,  back_genes)
inSelection <-  gene_IDs %in% DE_genes_CD 
all_genes <- factor(as.integer(inSelection[in_universe]))
names(all_genes) <- gene_IDs[in_universe]
```

## GO 注释
We now initialize the `r Biocpkg("topGO") ` data set, using the GO annotations 
contained
in the annotation data base for the chip we are using. The `nodeSize`
parameter specifies a minimum size of a GO category we want to use: i.e. here
categories with less than 10 genes are not included in the testing.

```{r createTopGODataSet, dependson="createFactorOfInterestingGenes", eval=TRUE, message = FALSE }
ont <- "BP"

top_GO_data <- new("topGOdata", ontology = ont, allGenes = all_genes,
 nodeSize = 10, annot=annFUN.db, affyLib = "hgu133a2.db")
```

Now the tests can be run. `r Biocpkg("topGO") ` offers a wide range of options,
for details see the paper or the package vignette.

We run two common tests: an ordinary Fisher test for every GO category, and the
"elim" algorithm, which tries to incorporate the hierarchical structure of the
GO and tries "decorrelate" it in order to report the most specific significant 
term in the hierarchy.

The algorithm starts processing the nodes/GO categories
from the highest (bottommost) level and then iteratively
moves to nodes from a lower level. If a node is scored as significant,
all of its genes  are marked as removed in all ancestor nodes.
This way, the "elim" algorithm aims at finding the most specific node
for every gene.

The tests uses a 0.01 p--value cutoff by default.

## GO 富集分析检验

```{r runtopGOTests, results='hide', eval=TRUE, dependson = "createTopGODataSet",  message = FALSE}
result_top_GO_elim <- runTest(top_GO_data, algorithm = "elim", statistic = "Fisher")
result_top_GO_classic <- runTest(top_GO_data, algorithm = "classic", statistic = "Fisher")
```

We can now inspect the results.  We look at the top 100 GO categories according
to the "Fisher elim" algorithm. The function `GenTable` produces
a table of significant GO categories, the function `printGenes`
gives significant genes annotated to them.

```{r processtopGOResults, eval=TRUE, dependson="runtopGOTests"}
res_top_GO <- GenTable(top_GO_data, Fisher.elim = result_top_GO_elim,
        Fisher.classic = result_top_GO_classic,
        orderBy = "Fisher.elim" , topNodes = 100)

genes_top_GO <- printGenes(top_GO_data, whichTerms = res_top_GO$GO.ID,
    chip = "hgu133a2.db", geneCutOff = 1000)

res_top_GO$sig_genes <- sapply(genes_top_GO, function(x){
                str_c(paste0(x[x$'raw p-value' == 2, "Symbol.id"],";"), collapse = "")
    })

head(res_top_GO[,1:8], 20)
```

## 可视化富集结果
Visualization of the GO--analysis results

A graph of the results can also be produced. Here we visualize the three most
significant nodes according to the Fisher elim algorithm in the context of
the GO hierarchy.

```{r  graph_of_results, fig.height = 6, eval=TRUE, results='hide'}
showSigOfNodes(top_GO_data, score(result_top_GO_elim), firstSigNodes = 3,
               useInfo = 'def')
```

We can see that indeed GO categories related to inflammation, signalling and 
immune response show up as significant.
Gene set enrichment analysis has been  a field of very extensive
research in bioinformatics. For additional approaches see the `r Biocpkg("topGO") `
vignette and the references therein and also in the [GeneSetEnrichment view](http://bioconductor.org/packages/release/BiocViews.html#___GeneSetEnrichment). 


# Reactome pathway 富集分析
A pathway enrichment analysis using reactome

The package `r Biocpkg("ReactomePA") ` offers the possibility to test enrichment
of specific pathways using the free, open-source, curated and peer reviewed 
pathway [Reactome](http://www.reactome.org/s) pathway database [@Croft_2013; @Fabregat_2015]. 
The package requires entrez identifiers, so we convert our PROBEIDs (trancript cluster identifiers) 
to entrez identifiers using the function `mapIDs` from  the package `r Biocpkg("AnnotationDbi")`. 
This will create a named vector that maps the PROBEIDs to the entrez ones.


```{r mapIDsToEntrez, dependson="createFactorOfInterestingGenes", message = FALSE}
entrez_ids <- mapIds(hgu133a2.db, 
      keys = rownames(table_CD), 
      keytype="PROBEID",
      column = "ENTREZID")
```


We can now run the enrichment analysis that performs a statistical test
based on the hypergeoemtric distribution that is the same as a one sided Fisher--test, 
which `r Biocpkg("topGO")` calls "Fisher--classic".
Details can be found in the vignette of the `r Biocpkg("DOSE")` package [@Yu_2014].

```{r runReactomeEnrichment, dependson="mapIDsToEntrez", eval = TRUE, cache=TRUE}
library(ReactomePA)
reactome_enrich <- enrichPathway(gene = entrez_ids[DE_genes_CD], 
                                universe = entrez_ids[c(DE_genes_CD, 
                                                        back_genes)],
                                organism = "human",
                                pvalueCutoff = 0.05,
                                qvalueCutoff = 0.9, 
                                readable = TRUE)

#reactome_enrich@result$Description <- paste0(str_sub(
#                                    reactome_enrich@result$Description, 1, 100),
#                                   "...")

head(summary(reactome_enrich))[1:6]
```

Note that we trimmed pathway names to 20 characters.

## 富集结果可视化
Visualizing the reactome based analysis results 

The `r Biocpkg("reactomePA") ` package offers nice visualization capabilities.
The top pathways can be displayed as a bar char that displays all categories
with a p--value below the specified cutoff.

##条形图
```{r reactomeBar, dependson="runReactomeEnrichment", eval = TRUE}
barplot(reactome_enrich)
```

The "enrichment map" displays the results of the enrichment analysis as 
a graph, where the color represents the p--value of the pathway and the
edge--thickness is proportional to the number of overlapping genes between
two pathways.

##富集项的网络关联

```{r enrichMap, dependson="runReactomeEnrichment", fig.width=6, fig.height = 7, eval = TRUE}
enrichMap(reactome_enrich, n = 10, vertex.label.font = 2)
```

Again, we see pathways related to  signalling and immune response. 

The package `r Biocpkg("clusterProfiler") ` [@Yu_2012] can also perform these analyses 
using downloaded KEGG data. Furthermore, the package `r Biocpkg("EnrichmentBrowser") `
[@Geistlinger_2016] additionally offers network--based enrichment analysis of individual 
pathways. This allows the mapping of the expression data at hand to known
regulatory interactions.


```{r}
sessionInfo()
```



<!--
Dan Tenenbaum provided helpful technical support on the bioconductor
workflow system. 
-->



